Mathlib/FieldTheory/SeparableDegree.lean

```diff
@@ -268,31 +249,539 @@ def embProdEmbOfIsAlgebraic [Algebra E K] [IsScalarTower F E K] (halg : Algebra.
-    sepDegree A B * sepDegree B C = sepDegree A C := by
-  simpa only [Cardinal.lift_id] using lift_sepDegree_mul_lift_sepDegree_of_isAlgebraic A B C halg
-
-/-- The `Field.finSepDegree` version of `Field.lift_sepDegree_mul_lift_sepDegree_of_isAlgebraic`.
-See also `FiniteDimensional.finrank_mul_finrank'`. -/
+/-- If `K / E / F` is a field extension tower, such that `K / E` is algebraic, then their
+(finite) separable degree satisfies the tower law
+$[E:F]_s [K:E]_s = [K:F]_s$. See also `FiniteDimensional.finrank_mul_finrank'`. -/
 theorem finSepDegree_mul_finSepDegree_of_isAlgebraic
     [Algebra E K] [IsScalarTower F E K] (halg : Algebra.IsAlgebraic E K) :
     finSepDegree F E * finSepDegree E K = finSepDegree F K := by
-  simpa only [Cardinal.toNat_mul, Cardinal.toNat_lift] using congr_arg Cardinal.toNat
-    (lift_sepDegree_mul_lift_sepDegree_of_isAlgebraic F E K halg)
+  simpa only [Nat.card_prod] using Nat.card_congr (embProdEmbOfIsAlgebraic F E K halg)
+
+end Field
+
+namespace Polynomial
+
+variable {F E}
+
+variable (f : F[X])
+
+/-- If a polynomial has all roots over a field, then it has no repeated roots on that field
+if and only if it is separable. -/
+theorem nodup_roots_iff_of_splits (hf : f ≠ 0) (h : f.Splits (RingHom.id F)) :
+    f.roots.Nodup ↔ f.Separable := by
+  refine ⟨not_imp_not.1 fun hnsep ↦ ?_, nodup_roots⟩
+  set g := gcd f (derivative f)
+  have hg : g ≠ 0 := gcd_ne_zero_of_left hf
+  by_cases hdeg : g.degree = 0
+  · simp only [Separable, ← gcd_isUnit_iff, isUnit_iff] at hnsep
+    rw [eq_C_of_degree_eq_zero hdeg] at hg
+    exact False.elim <| hnsep ⟨coeff g 0, isUnit_iff_ne_zero.2 (ne_zero_of_map hg),
+      (eq_C_of_degree_eq_zero hdeg).symm⟩
+  obtain ⟨x, hx⟩ := exists_root_of_splits _
+    (splits_of_splits_of_dvd _ hf h (show g ∣ f from gcd_dvd_left f _)) hdeg
+  rw [Multiset.nodup_iff_count_le_one, not_forall]
+  exact ⟨x, by rw [count_roots]; exact Nat.not_le.mpr <|
+    (one_lt_rootMultiplicity_iff_isRoot_gcd hf).2 hx⟩
+
+/-- The separable degree `Polynomial.natSepDegree` of a polynomial is a natural number,
+defined to be the number of distinct roots of it over its splitting field.
+This is similar to `Polynomial.natDegree` but not `Polynomial.degree`, namely, the separable
+degree of `0` is `0`, but not negative infinity. -/
+def natSepDegree : ℕ := (f.aroots f.SplittingField).toFinset.card
+
+/-- The separable degree of a polynomial is smaller than its degree. -/
+theorem natSepDegree_le_natDegree : f.natSepDegree ≤ f.natDegree := by
+  have := f.map (algebraMap F f.SplittingField) |>.card_roots'
+  rw [← aroots_def, natDegree_map] at this
+  exact (f.aroots f.SplittingField).toFinset_card_le.trans this
+
+@[simp]
+theorem natSepDegree_X_sub_C {x : F} : (X - C x).natSepDegree = 1 := by
+  simp only [natSepDegree, aroots_X_sub_C, Multiset.toFinset_singleton, Finset.card_singleton]
+
+@[simp]
+theorem natSepDegree_X : (X : F[X]).natSepDegree = 1 := by
```

Comment:
Here and in all the following lemma where it is not possible to infer it, `F` should be explicit.

Reply:
Oh but in [Polynomial.natDegree_X_le](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/Polynomial/Degree/Definitions.html#Polynomial.natDegree_X_le) the ring is implicit; I just mimic the pattern of the existing API.